### PR TITLE
Add a small test app for generating secure URLs

### PR DESCRIPTION
**Pull request by jon-betts**
This helps with generating signed URLs for Via 3 and Via HTML.

 * Runs in SSL (self signed)
 * Can sign a single, or multiple URLs
 * If a single URL is used, it will show it in an iframe
 * Supports Via 3 and Via HTML in dev, QA and prod

## Testing notes

* `cd app/via_test_harness`
 * Add the relevant environment secrets to `conf/secrets.json`
 * Run: `make dev`
 * Visit: https://0.0.0.0:9101/
 * This service runs on SSL with a self signed certificate, so you will need
  to allow it in Chrome
 * You can disable this with `conf/config.json: ssl` option

### CODE REVIEW

File: app/via_test_harness/Makefile [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+	@echo "make help	Show this help message"
+	@echo "make dev		Run the app in the development server"
+
+
+.PHONY: dev
+dev: venv
+	@.venv/bin/python app.py
+
+.PHONY: venv
+venv: .venv/bin/pip
```
Comment (jon-betts) [inline app/via_test_harness/Makefile:12]: I expected this to prevent rebuilds every time, but it doesn't work. It's quite quick anyway

**Review by seanh — CHANGES_REQUESTED** [review]
If this is going to be merged then it becomes the whole team's burden to maintain, so I'd ask you to please change it to Pyramid rather than Flask so that we don't have to deal with two web frameworks unnecessarily. It's very simple to write a small Pyramid app, see the 10-line hello world app at [the start of their docs](https://docs.pylonsproject.org/projects/pyramid/en/latest/) or the one I made for [testing PDF redirects](https://github.com/hypothesis/pdf_redriects/blob/main/pdf_redirects/app.py#L1-L47).

I'm not sure about the SSL part but we have run Gunicorn + Pyramid apps locally with SSL before. For example h has `make devssl`. 

I'll also ask for it to use our usual tox-based virtualenv managements. This can also be copied from [the pdf_redirects app](https://github.com/hypothesis/pdf_redriects)

**Comment by jon-betts** [comment]
This whole app is less than 100 lines long, of which maybe 12 are Flask specific. The example provided is over 5 times the size as a result of following all the standard conventions.

I don't think that amount of work is appropriate for this kind of dev tool. If there is anything we cannot understand or fix, and it's genuinely easy to write again as a Pyramid app, we can do it if we hit a problem we can't solve.

I don't know of any quick solution to serving an app over SSL with Pyramid as provided here.

**Comment by robertknight** [comment]
Following the instructions above to run this, I encountered this error when running `make dev`:

```
$ make dev
make: *** No rule to make target `.venv/bin/pip', needed by `venv'.  Stop.
```

I fixed this locally by changing the Makefile to adopt a pattern that we use in the client repo (and I think a few others) to automatically run the package manager when needed:

```diff
diff --git a/app/via_test_harness/Makefile b/app/via_test_harness/Makefile
index ad55cb7..05abe32 100644
--- a/app/via_test_harness/Makefile
+++ b/app/via_test_harness/Makefile
@@ -5,10 +5,10 @@ help:
 
 
 .PHONY: dev
-dev: venv
+dev: .venv/uptodate
 	@.venv/bin/python app.py
 
-.PHONY: venv
-venv: .venv/bin/pip
+.venv/uptodate: requirements.txt
 	@virtualenv .venv -p python3
 	@.venv/bin/pip install -r requirements.txt
+	@touch .venv/uptodate
```

**Comment by seanh** [comment]
I do have a slippery-slope-ish concern here, yes. I don't think my concern is _quite_ the same thing as a slippery-slope argument though. I don't think the existence of a test app in Flask is going to lead to us unnecessarily using Flask in production. It's more of a broken windows argument. The effect on our team's mindset of a generally clean and consistent codebase and practices across all our projects, versus a messy and variable one. Especially when thinking of new recruits and expanding the team.

Regarding the amount of time that it takes to add all the tooling etc just for a small project: I totally get this, and would like to address it. A couple of ideas I have: I'd like to expand `qt` to become an `hdev` command line app with commands like `hdev test`, `hdev lint`, `hdev format`, etc etc. Implementing those commands in one centralised, installable script, and then removing all the stuff needed for `make test`, `make lint`, `make format` etc from all the individual projects. For what's still left behind in the projects, I'd like to write an app cookiecutter so you can spit it all out in one command and just get coding.

I'd really appreciate if you could help me out by keeping things consistent. I've put a lot of time and effort into making our once totally inconsistent suite of projects all consistent, I think it makes a big difference and the team has really appreciated it. Test tools are software too: they still need to be developed and maintained by our team. We do have a practice of creating test / demo apps using Pyramid (where a web framework is relevant) and our usual dev env setup. For example: the [pdf_redirects](https://github.com/hypothesis/pdf_redriects) app for testing Via with redirecting PDFs, [potemkin](https://github.com/hypothesis/potemkin) for easier testing of the LMS app without an actual LMS, and others.

We usually put these test apps each in its own repo too. I've changed this to Pyramid and our usual dev env stuff. Give this a try: https://github.com/hypothesis/via_test_app I had to change how the SSL support works to make it work with Gunicorn / Pyramid, which means that same solution (after fixing any mistakes I made or anything that I missed) could be applied to the rest of our apps.

FYI: There is already a repo that you previously created called via_test_harness! https://github.com/hypothesis/via_test_harness

Also FYI: We actually have an app for easier local testing of the LMS app (including Via 3 and Via HTML as called by the LMS app: https://github.com/hypothesis/potemkin That simulates an actual LMS environment, doing an LTI launch in an iframe and showing the LMS app / Via's response. In its current form you can't easily use it to test lots of different HTML URLs like this new Via test app is intended to do. But we might want to consider extending Potemkin to add more functionality.

**Comment by seanh** [comment]
@robertknight Let's just use tox to keep Python dependencies up to date! I'd like to avoid new solutions like `Makefile` code that touches a `.venv/uptodate` file as now we've solved the same problem in two different ways and tooling has begun to diverge. https://github.com/hypothesis/via_test_app uses tox to manage the virtualenvs in the same way as all our other projects do

**Comment by robertknight** [comment]
The ad-hoc SSL certificate provisioned by Flask doesn't appear to work in Safari. I ended up in the following loop:

Step 1:

Safari SSL cert step 1

Step 2:

Safari SSL cert step 2

Step 3:

Safari SSL cert step 3

At this point clicking "Visit website" prompts me for my macOS user password (or fingerprint). This adds the cert to the system store, but with no useful information about what it was for:

adhoc keychain

I then get taken back to Step 1. 

Getting SSL working in general is unfortunately fiddly. I wrote up some notes on how I got it working with h and the LMS app in the past here: https://hypothes-is.slack.com/archives/DPPMK856Z/p1589477985006200. One addendum to that Slack thread is that Chrome _does_ not treat http://localhost as secure, although I'm not sure if that applies to every single thing in the browser that differentiates between secure and non-secure origins. Hence in Chrome using SSL is no longer worthwhile any more.

**Comment by jon-betts** [comment]
This has been superceded

**Pull request closed without merging**